WWDC/DownloadManager.swift

```diff
@@ -87,96 +79,60 @@ final class DownloadManager: NSObject {
         monitorDownloadsFolder()
     }
 
-    fileprivate func localStoragePath(for asset: SessionAsset) -> String {
-        return Preferences.shared.localVideoStorageURL.appendingPathComponent(asset.relativeLocalURL).path
-    }
+    func download(_ sessions: [Session]) {
+        var sessionURLMap = [SessionIdentifier: String]()
+        for session in sessions {
+            guard let asset = session.asset(ofType: DownloadManager.downloadQuality) else { continue }
 
-    func allTasks() -> [URLSessionDownloadTask] {
-        return Array(downloadTasks.values)
-    }
+            let url = asset.remoteURL
 
-    func download(_ asset: SessionAsset) {
-        downloadRemoteUrl(asset.remoteURL)
-    }
+            if isDownloading(url) || hasDownloadedVideo(asset: asset) {
+                continue
+            }
 
-    private func downloadRemoteUrl(_ url: String) {
-        if isDownloading(url) || hasVideo(url) {
-            return
+            sessionURLMap[SessionIdentifier(session.identifier)] = url
         }
 
-        let task = backgroundSession.downloadTask(with: URL(string: url)!)
-        if let key = task.originalRequest?.url!.absoluteString {
-            downloadTasks[key] = task
-        }
-        task.resume()
+        DispatchQueue.global(qos: .background).async {
+            var successfullyStartedTasks = [String: Download]()
+            for (sessionID, url) in sessionURLMap {
+                let task = self.backgroundSession.downloadTask(with: URL(string: url)!)
+                if let key = task.originalRequest?.url!.absoluteString {
+                    successfullyStartedTasks[key] = Download(session: sessionID, remoteURL: key, task: task)
+                } else {
+                    NotificationCenter.default.post(name: .DownloadManagerDownloadFailed, object: url)
+                }
+            }
 
-        NotificationCenter.default.post(name: .DownloadManagerDownloadStarted, object: url)
-    }
+            DispatchQueue.main.async {
+                self.downloadTasks.merge(successfullyStartedTasks, uniquingKeysWith: { a, b in b })
 
-    func pauseDownload(_ url: String) -> Bool {
-        if let task = downloadTasks[url] {
-            task.suspend()
-            NotificationCenter.default.post(name: .DownloadManagerDownloadPaused, object: url)
-            return true
+                for (url, download) in successfullyStartedTasks {
+                    download.task?.resume()
+                    NotificationCenter.default.post(name: .DownloadManagerDownloadStarted, object: url)
+                }
+            }
```

Comment:
It's not clear to me what the code above is doing, should probably add some comments.